zapcore/sampler.go

```diff
@@ -126,9 +193,36 @@ func (s *sampler) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
 	}
 
 	counter := s.counts.get(ent.Level, ent.Message)
-	n := counter.IncCheckReset(ent.Time, s.tick)
+	n, prevN := counter.IncCheckReset(ent.Time, s.tick)
 	if n > s.first && (n-s.first)%s.thereafter != 0 {
 		return ce
 	}
+	if s.reporting.enabled && prevN > 0 {
+		s.report(ent, prevN)
+	}
 	return s.Core.Check(ent, ce)
 }
+
+func (s *sampler) report(ent Entry, n uint64) error {
+	sampleCount := n - s.first
+	sampleCount = sampleCount - sampleCount/s.thereafter
+	if sampleCount <= 0 {
+		return nil
+	}
+	entry := Entry{
+		LoggerName: s.reporting.loggerName,
+		Time:       time.Now(),
+		Level:      s.reporting.level,
+		Message:    s.reporting.message,
+	}
+	fields := []Field{
+		{Key: "original_level", Type: StringType, String: ent.Level.String()},
+		{Key: "original_message", Type: StringType, String: ent.Message},
+		{Key: "count", Type: Int64Type, Integer: int64(sampleCount)},
+	}
+	err := s.reporting.core.Write(entry, fields)
```

Comment:
nit: just `return s.reporting.core.Write(...)`; you could even inline the fields in the log site; clearer to read if you use field constructors, also less fragile if we ever iterate on field particulars.

Reply:
Updated return.
To use field constructors here would require importing zap, which would introduce circular reference.
Not sure what you mean by 'inline the fields in the log site'. If you're talking about construction like:
```
reportingCore := s.reporting.core.With(fields)
reportingCore.Write()
```
I don't think it makes it any simpler or neater.